### ISSUE TITLE
[BUG] Auto-detected date is day before receipt date

### ISSUE THREAD

**Issue by minideezel** [paperless-ngx/paperless-ngx#224]
**Describe the bug**
When the system automatically detects the date on the receipt, the "Date Created" field is set to the day before.

**To Reproduce**

1. Have system timezone set to UTC-6
2. Scan document that has approved date type
3. See that date is day before date in receipt.

**Expected behavior**
Date should be exact date on receipt.

**Webserver logs**
N/A

**Relevant information**
 - Host OS of the machine running paperless: Docker/Ubuntu 18.04
 - Browser: Chrome
 - Version 1.5.0
 - Installation method: docker
 - Any configuration changes:
```
PAPERLESS_TIME_ZONE=America/Chicago
PAPERLESS_DATE_ORDER=MDY
```

**Comment by qcasey** [comment]
Thanks for the report. Possibly related:

https://github.com/jonaswinkler/paperless-ng/issues/1301
https://github.com/jonaswinkler/paperless-ng/issues/1417
https://github.com/jonaswinkler/paperless-ng/issues/1344
https://github.com/jonaswinkler/paperless-ng/issues/872

**Comment by torwag** [comment]
I tried to have a look into it (somewhere one has to start, and this bug seems to be both... annoying enough and "easy" enough),
First of, for some reason code search on github for paperless-ngx seems not to work, have no idea why.

On the original paperless-ng, the following line of code might be the starter
[date-selection](https://github.com/jonaswinkler/paperless-ng/blob/05c36f91cfa5613bdea5b3c490774cc40b586eeb/src/documents/consumer.py#L376)
Seems it prefers first dates fetched from the file name, then dates obtained by OCR and if this is not possible it falls back to the current date at time of consumption.

My question now (I know it is rather unlikely), could it be that 

a) the file name has a date  that is for whatever reason exactly one day before?
b) did you scan the document on a system which has not set the right timezone?

That would exclude that the parser really just does what it should and that the input data is simply wrong.

If you say, the "scan document that has approved date type" you mean the data in the document fetched by OCR, right?

**Comment by torwag** [comment]
This was the answer of [btorresgil](https://github.com/btorresgil) which seems to be much more on the right spot already, copy here verbatim, to keep things in paperless-ngx (in case paperless-ng is down somehow in the future)

This issue seems to be caused by the usage of dateparser.parse() here:

[paperless-ng/src/documents/parsers.py](https://github.com/jonaswinkler/paperless-ng/blob/7bc8325df910ab57ed07849a3ce49a3011ba55b6/src/documents/parsers.py#L221)

Line 221 in [7bc8325](https://github.com/jonaswinkler/paperless-ng/commit/7bc8325df910ab57ed07849a3ce49a3011ba55b6)
 return dateparser.parse( 

On most scanned documents, there will be a date without a time or timezone. dateparser is configured to return a timezone aware datetime, even if no time or timezone is found in the document. However, it does not acknowledge the timezone set by PAPERLESS_TIME_ZONE. So when it finds a date with no time or timezone (the most common use case) it creates a datetime with the time set to 00:00:00 and the timezone as the server system timezone (UTC by default for docker containers).

Let's take an example where a document has the date Aug 11 2021 at the top of the document when it is scanned.

This is stored in the database as: 2021-08-11 00:00:00+00.

The problem comes in when this datetime is read later, as before it is displayed to the user it is converted to the PAPERLESS_TIME_ZONE timezone. So for someone in America/Los_Angeles this is sent to the frontend as 2021-08-10T17:00:00-07:00, which is incorrectly displayed by the frontend as 08/10/2021.
Workaround

Since dateparser uses the system timezone when no timezone is found in the document text, a workaround is to set the TZ environment variable to change the system timezone in the docker container. So for example, if you're using the docker-compose.env file, you'd change it to include a TZ environment variable:

# Use this variable to set a timezone for the Paperless Docker containers. If not specified, defaults to UTC.
PAPERLESS_TIME_ZONE=America/Los_Angeles
TZ=America/Los_Angeles                         <---------- Add this

Unfortunately, this doesn't fix any documents already stored because it only affects the timezone used in the consumption step. All future documents will get a created fields that is adjusted for timezone, so in the example above the database will store 2021-08-11 07:00:00+0 (instead of 2021-08-11 00:00:00+0) which will be correctly interpreted by the frontend as 08/11/2021.

(Workaround is also related to https://github.com/jonaswinkler/paperless-ng/issues/872)
Possible fixes

One possible fix is to stop storing time/timezone at all, or stop sending time/timezone to the frontend, since the frontend doesn't show the time anyway, only the date. Of course, this would be a breaking change as the API would respond without time or timezone.

Anther possible fix, perhaps it would be helpful to add a setting to dateparser.parse() that sets the TIMEZONE setting to the PAPERLESS_TIME_ZONE. I haven't tested this, but the docs seem to indicate it would work:

from: https://pypi.org/project/dateparser/
Example where TIMEZONE is used but no timezone in date string, string is interpreted as being in specified timezone:

>>> parse('January 12, 2012 10:00 PM', settings={'TIMEZONE': 'US/Eastern', 'RETURN_AS_TIMEZONE_AWARE': True})
datetime.datetime(2012, 1, 12, 22, 0, tzinfo=<DstTzInfo 'US/Eastern' EST-1 day, 19:00:00 STD>)

    In case, when timezone is present both in string and also specified using settings, string is parsed into tzaware representation and then converted to timezone specified in settings.

**Comment by torwag** [comment]
I would prefer
if no timezone is set in the document/string (which is as said pretty much standard)
1. set PAPERLESS_TIME_ZONE in the settings of dateparser.parse.
2. PAPERLESS_TIME_ZONE is not set, use TZ as default not UTC
3. in the provided env files, make sure that this is clear:
`# Use this variable to set a timezone for the Paperless-NGX. If not specified, it defaults to TZ which is usually set to UTC in a docker container by default.
#PAPERLESS_TIME_ZONE=America/Los_Angeles`
4. Add documentation particular in the docker set-up that PAPERLESS_TIME_ZONE is set to TZ if not set explicitly and that TZ falls back to UTC

Why adding this chain? If people modify PAPERLESS_TIME_ZONE explicitly, they expect this to be the timezone. If nothing is set, it needs to be clear, that a docker container uses UTC. As this is a common misconception not only for paperless-ngx but for many docker containers, people will learn to set TZ in the docker-compose file to be the timezone of the host. For some people, this becomes a habit. This common solution should work with paperless-ngx too.

**Comment by a17t** [comment]
@torwag please look into the linked PR

**Comment by torwag** [comment]
@a17t sorry for some reasons I couldn't see the linked PR. Great!
I would add the change in the docker env file to drop back to TZ to make sure people can modify TZ to achive the same.

**Comment by qcasey** [comment]
Fixed in branch #246
